File: src/main/java/org/poker/irc/messagehandler/StreamsMessageEventHandler.java
package org.poker.irc.messagehandler;

import org.javatuples.Pair;
import org.pircbotx.hooks.events.MessageEvent;
import org.poker.irc.Configuration;
import org.poker.irc.HttpUtils;
import org.poker.irc.MessageEventHandler;
import org.poker.irc.twitch.StreamsResponse;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import sun.reflect.generics.reflectiveObjects.NotImplementedException;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;

public class StreamsMessageEventHandler implements MessageEventHandler {
  private enum Game {
    Dota,
    LeagueOfLegends,
    Quake,
    MagicTheGathering,
    Diablo3,
    Rust,
    Hearthstone,
  }
  private static final Logger LOG = LoggerFactory.getLogger(StreamsMessageEventHandler.class);
  private final Configuration configuration;

  public StreamsMessageEventHandler(Configuration configuration) {
    this.configuration = configuration;
  }

  @Override
  public String getMessageRegex() {
    return null;
  }

  @Override
  public String getDescription() {
    return "!streams <game> or .streams <game>: send to channel top twitch streams for { dota, lol, quake, magic }.";
  }

  @Override
  public String[] getMessagePrefixes() {
    return new String[] { "!twitch", ".twitch", "!streams", ".streams" };
  }

  @Override
  public void onMessage(MessageEvent event) {
    Game game;
    String message = event.getMessage();
    int index = message.indexOf(' ');
    if (index < 0) {
      game = Game.Dota;
    } else {
      // TODO: switch to regex for this stuff
      String gameName = message.substring(index + 1).trim().toLowerCase();
      if (gameName.startsWith("l")) {
        game = Game.LeagueOfLegends;
      } else if (gameName.startsWith("do")) {
        game = Game.Dota;
      } else if (gameName.startsWith("di")) {
        game = Game.Diablo3;
      } else if (gameName.startsWith("q")) {
        game = Game.Quake;
      } else if (gameName.startsWith("mtg") || gameName.startsWith("magic") || gameName.startsWith("m:tg")) {
        game = Game.MagicTheGathering;
      } else if (gameName.startsWith("rust")) {
        game = Game.Rust;
      } else if (gameName.startsWith("hearth")) {
        game = Game.Hearthstone;
      }else if (gameName.equals("poker")) {
        event.getChannel().send().message("Poker? I don't support dead games");
        return;
      } else {
        event.getChannel().send().message("Unknown game: " + gameName);
        return;
      }
    }
    int limit = 3;
    String gameName;
    switch (game) {
      case Dota:
        gameName = "Dota+2";
        break;
      case LeagueOfLegends:
        gameName = "League+of+Legends";
        break;
      case MagicTheGathering:
        gameName = "Magic%3A%20The%20Gathering";
        break;
      case Quake:
        gameName = "Quake%20Live";
        limit = 1;
        break;
      case Rust:
        gameName = "Rust";
        limit = 2;
        break;
      case Hearthstone:
        gameName = "Hearthstone";
        break;
      case Diablo3:
        //StandardCharsets.UTF_8.displayName()
        try {
          gameName = URLEncoder.encode("Diablo III: Reaper of Souls", "ISO-8859-1");
        } catch (UnsupportedEncodingException e) {
          throw new RuntimeException(e);
        }
        limit = 3;
        break;
      default:
        throw new NotImplementedException();
    }
    String url = "https://api.twitch.tv/kraken/streams?limit=" + limit + "&game=" + gameName;
    StreamsResponse streamsResponse = HttpUtils.getJson(url, StreamsResponse.class,
        new Pair<String, String>("Client-ID", this.configuration.getTwitchClientId()),
        new Pair<String, String>("Accept", "application/vnd.twitchtv.v2+json"));
    for (StreamsResponse.Stream stream : streamsResponse.getStreams()) {
      StringBuilder sb = new StringBuilder();
      sb.append(stream.getChannel().getDisplay_name());
      sb.append(" | ");
      sb.append(stream.getViewers());
      sb.append(" viewers | ");
      sb.append(stream.getChannel().getUrl());
      sb.append("/popout");
      event.getChannel().send().message(sb.toString());
    }
  }
}

File: src/main/scala/org/poker/handler/StreamsMessageEventHandler.scala
package org.poker.handler

import org.poker.ProgramConfiguration
import org.poker.twitch.{Game, TwitchClient}
import sx.blah.discord.handle.impl.events.MessageReceivedEvent

import scala.util.matching.Regex
import scala.util.matching.Regex.Match

class StreamsMessageEventHandler(configuration: ProgramConfiguration) extends MessageEventHandler {
  val baseUrl = "https://api.twitch.tv/kraken/"
  val twitchClient = new TwitchClient(configuration.twitchClientId.getOrElse(""))

  override val helpMessage: Option[String] = Option("!streams <query>: send to channel the top twitch search results for <query>")

  override val messageMatchRegex: Regex = "^[!.](?i)streams? ?(?<query>.*)".r

  override def onMessage(event: MessageReceivedEvent, firstMatch: Match): Unit = {
    if (configuration.twitchClientId.isDefined) {
      val query = firstMatch.group(1)
      handleQuery(event, query)
    } else {
      event.getMessage.getChannel.sendMessage("twitch disabled: no client id")
    }
  }

  private def queryGames(query: String): List[Game] = {
    if (query.isEmpty) {
      twitchClient.queryGames("Dota 2").games
    } else {
      twitchClient.queryGames(query).games
    }
  }

  private def handleQuery(event: MessageReceivedEvent, query: String) {
    val games = this.queryGames(query)
    if (games.isEmpty) {
      event.getMessage.getChannel.sendMessage(s"no twitch games found for query: '$query'")
    } else {
      val game = games(0)
      val streamsResponse = twitchClient.queryStreams(game.name, 3)
      if (streamsResponse.streams.isEmpty) {
        event.getMessage.getChannel.sendMessage(s"sorry, game '${game.name}' is dead")
      } else {
        event.getMessage.getChannel.sendMessage(s"streams for '${game.name}'")
        for (stream <- streamsResponse.streams) {
          val message = s"${stream.channel.display_name} | ${stream.viewers} | ${stream.channel.url}/popout"
          event.getMessage.getChannel.sendMessage(message)
        }
      }
    }
  }
}
